File: backend/app/connectors/jira_connector.py
from typing import List, Dict, Any, Optional, Tuple
from datetime import datetime
from atlassian import Jira
from .base_connector import BaseConnector

class JiraConnector(BaseConnector):
    def __init__(self):
        super().__init__()
        self.display_name = "JIRA"
        self.description = "Atlassian JIRA issues and project documentation"
        self.jira = None
        self.credentials = {}
    
    def set_credentials(self, credentials: Dict[str, str]):
        """Set credentials for this connector"""
        self.credentials = credentials
        self._initialize()
    
    def _initialize(self):
        """Initialize JIRA client with provided credentials"""
        url = self.credentials.get("jira_url")
        username = self.credentials.get("jira_username")
        api_token = self.credentials.get("jira_api_token")
        
        if url and username and api_token:
            self.jira = Jira(
                url=url,
                username=username,
                password=api_token,
                cloud=True
            )
    
    async def is_configured(self) -> bool:
        """Check if JIRA is configured"""
        return self.jira is not None
    
    async def test_connection(self) -> bool:
        """Test connection to JIRA"""
        if not self.jira:
            return False
        
        try:
            self.jira.get_all_projects()
            return True
        except Exception:
            return False
    
    async def fetch_documents(self) -> List[Dict[str, Any]]:
        """Fetch issues and documentation from JIRA"""
        if not self.jira:
            return []
        
        documents = []
        
        try:
            # Get all projects first
            projects = self.jira.get_all_projects()
            project_keys = [p['key'] for p in projects[:5]]  # Limit to first 5 projects
            
            for project_key in project_keys:
                # Get issues from project
                jql = f"project = {project_key} ORDER BY updated DESC"
                issues = self.jira.jql(jql, limit=100, fields='*all')
                
                for issue in issues.get('issues', []):
                    doc = self._parse_issue(issue)
                    if doc:
                        documents.append(doc)
        
        except Exception as e:
            print(f"Error fetching JIRA documents: {e}")
        
        return documents
    
    def _parse_issue(self, issue: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Parse a JIRA issue into a document"""
        try:
            fields = issue.get('fields', {})
            
            # Build comprehensive content
            content_parts = []
            
            # Summary
            summary = fields.get('summary', '')
            if summary:
                content_parts.append(f"Summary: {summary}")
            
            # Description
            description = fields.get('description', '')
            if description:
                content_parts.append(f"Description: {description}")
            
            # Priority
            priority = fields.get('priority', {})
            if priority:
                content_parts.append(f"Priority: {priority.get('name', 'None')}")
            
            # Assignee
            assignee = fields.get('assignee', {})
            if assignee:
                content_parts.append(f"Assignee: {assignee.get('displayName', 'Unassigned')}")
            
            # Labels
            labels = fields.get('labels', [])
            if labels:
                content_parts.append(f"Labels: {', '.join(labels)}")
            
            # Comments
            comments = fields.get('comment', {}).get('comments', [])
            comment_list = []
            for comment in comments:
                author = comment.get('author', {}).get('displayName', 'Unknown')
                body = comment.get('body', '')
                if body:
                    comment_list.append({
                        'author': author,
                        'body': body,
                        'created': comment.get('created', '')
                    })
            
            content = "\n\n".join(content_parts)
            
            if not content.strip():
                return None
            
            return {
                'id': issue['key'],
                'title': f"{issue['key']}: {summary}",
                'content': content,
                'description': description,
                'comments': comment_list,
                'url': f"{self.credentials.get('jira_url')}/browse/{issue['key']}",
                'created_at': fields.get('created', ''),
                'updated_at': fields.get('updated', ''),
                'issue_type': fields.get('issuetype', {}).get('name', ''),
                'status': fields.get('status', {}).get('name', ''),
                'priority': priority.get('name', 'None') if priority else 'None',
                'assignee': assignee.get('displayName', 'Unassigned') if assignee else 'Unassigned',
                'labels': labels
            }
        except Exception as e:
            print(f"Error parsing JIRA issue: {e}")
            return None
    
    async def fetch_documents_since(self, since_date: Optional[datetime] = None, sync_token: Optional[str] = None) -> Tuple[List[Dict[str, Any]], Optional[str]]:
        """Fetch issues modified since a specific date (incremental sync)"""

        
        if not self.jira:
            return [], None
        
        documents = []
        
        try:
            # Build JQL query for issues updated since date
            if since_date:
                date_str = since_date.strftime('%Y-%m-%d %H:%M')
                jql = f"updated >= '{date_str}' ORDER BY updated DESC"
            else:
                jql = "ORDER BY updated DESC"
            
            # Get issues with pagination support
            start_at = int(sync_token) if sync_token else 0
            max_results = 100
            
            issues = self.jira.jql(jql, limit=max_results, start=start_at, fields='*all')
            
            for issue in issues.get('issues', []):
                doc = self._parse_issue(issue)
                if doc:
                    documents.append(doc)
            
            # Calculate next token for pagination
            next_token = None
            if len(issues['issues']) == max_results:
                next_token = str(start_at + max_results)
            
            return documents, next_token
            
        except Exception as e:
            print(f"Error fetching incremental JIRA documents: {e}")
            return [], None

File: backend/src/main/java/com/intelligence/assistant/connector/JiraConnector.java
package com.intelligence.assistant.connector;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import okhttp3.*;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

/**
 * Connector for Atlassian JIRA
 */
@Component
public class JiraConnector extends BaseConnector {
    
    private final OkHttpClient httpClient;
    private final ObjectMapper objectMapper;
    private String baseUrl;
    
    public JiraConnector() {
        this.displayName = "JIRA";
        this.description = "Atlassian JIRA issues and project documentation";
        this.httpClient = new OkHttpClient();
        this.objectMapper = new ObjectMapper();
    }
    
    @Override
    public boolean isConfigured() {
        return credentials != null &&
                credentials.containsKey("jira_url") &&
                credentials.containsKey("jira_username") &&
                credentials.containsKey("jira_api_token");
    }
    
    @Override
    public boolean testConnection() {
        if (!isConfigured()) {
            return false;
        }
        
        try {
            baseUrl = credentials.get("jira_url");
            String url = baseUrl + "/rest/api/3/myself";
            
            Request request = buildRequest(url);
            Response response = httpClient.newCall(request).execute();
            
            return response.isSuccessful();
        } catch (Exception e) {
            logError("Connection test failed", e);
            return false;
        }
    }
    
    @Override
    public List<Map<String, Object>> fetchDocuments() {
        logActivity("Fetching all issues");
        List<Map<String, Object>> documents = new ArrayList<>();
        
        try {
            baseUrl = credentials.get("jira_url");
            
            // Fetch all issues with JQL
            String jql = "project is not EMPTY ORDER BY updated DESC";
            documents.addAll(fetchIssuesWithJQL(jql, 100));
            
            logActivity("Fetched " + documents.size() + " issues");
        } catch (Exception e) {
            logError("Error fetching documents", e);
        }
        
        return documents;
    }
    
    @Override
    public List<Map<String, Object>> fetchDocumentsSince(LocalDateTime sinceDate, String syncToken) {
        logActivity("Fetching issues since " + sinceDate);
        List<Map<String, Object>> documents = new ArrayList<>();
        
        try {
            baseUrl = credentials.get("jira_url");
            
            // Build JQL for incremental sync
            String dateStr = sinceDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
            String jql = String.format("updated >= '%s' ORDER BY updated DESC", dateStr);
            
            documents.addAll(fetchIssuesWithJQL(jql, 100));
            
            logActivity("Fetched " + documents.size() + " updated issues");
        } catch (Exception e) {
            logError("Error fetching incremental documents", e);
        }
        
        return documents;
    }
    
    private List<Map<String, Object>> fetchIssuesWithJQL(String jql, int maxResults) throws Exception {
        List<Map<String, Object>> issues = new ArrayList<>();
        
        String url = baseUrl + "/rest/api/3/search?jql=" + 
                java.net.URLEncoder.encode(jql, "UTF-8") + 
                "&maxResults=" + maxResults +
                "&fields=*all";
        
        Request request = buildRequest(url);
        Response response = httpClient.newCall(request).execute();
        
        if (response.isSuccessful() && response.body() != null) {
            String responseBody = response.body().string();
            JsonNode root = objectMapper.readTree(responseBody);
            JsonNode issuesNode = root.get("issues");
            
            if (issuesNode != null && issuesNode.isArray()) {
                for (JsonNode issue : issuesNode) {
                    issues.add(parseIssue(issue));
                }
            }
        }
        
        return issues;
    }
    
    private Map<String, Object> parseIssue(JsonNode issue) {
        Map<String, Object> document = new HashMap<>();
        
        String key = issue.get("key").asText();
        JsonNode fields = issue.get("fields");
        
        document.put("id", key);
        document.put("title", key + ": " + fields.get("summary").asText());
        
        // Build content
        StringBuilder content = new StringBuilder();
        content.append("Summary: ").append(fields.get("summary").asText()).append("\n\n");
        
        if (fields.has("description") && !fields.get("description").isNull()) {
            content.append("Description: ").append(fields.get("description").asText()).append("\n\n");
        }
        
        // Add priority
        if (fields.has("priority") && !fields.get("priority").isNull()) {
            content.append("Priority: ").append(fields.get("priority").get("name").asText()).append("\n");
        }
        
        // Add assignee
        if (fields.has("assignee") && !fields.get("assignee").isNull()) {
            content.append("Assignee: ").append(fields.get("assignee").get("displayName").asText()).append("\n");
        }
        
        // Add labels
        if (fields.has("labels") && fields.get("labels").isArray()) {
            List<String> labels = new ArrayList<>();
            for (JsonNode label : fields.get("labels")) {
                labels.add(label.asText());
            }
            if (!labels.isEmpty()) {
                content.append("Labels: ").append(String.join(", ", labels)).append("\n");
            }
        }
        
        document.put("content", content.toString());
        
        // Extract comments
        List<Map<String, String>> comments = new ArrayList<>();
        if (fields.has("comment") && fields.get("comment").has("comments")) {
            JsonNode commentsNode = fields.get("comment").get("comments");
            for (JsonNode comment : commentsNode) {
                Map<String, String> commentMap = new HashMap<>();
                commentMap.put("author", comment.get("author").get("displayName").asText());
                commentMap.put("body", comment.get("body").asText());
                commentMap.put("created", comment.get("created").asText());
                comments.add(commentMap);
            }
        }
        document.put("comments", comments);
        
        // Metadata
        document.put("url", baseUrl + "/browse/" + key);
        document.put("type", "jira_issue");
        document.put("issue_type", fields.get("issuetype").get("name").asText());
        document.put("status", fields.get("status").get("name").asText());
        document.put("project", fields.get("project").get("key").asText());
        document.put("project_name", fields.get("project").get("name").asText());
        document.put("created_at", fields.get("created").asText());
        document.put("updated_at", fields.get("updated").asText());
        
        return document;
    }
    
    private Request buildRequest(String url) {
        String username = credentials.get("jira_username");
        String apiToken = credentials.get("jira_api_token");
        String credential = Credentials.basic(username, apiToken);
        
        return new Request.Builder()
                .url(url)
                .header("Authorization", credential)
                .header("Accept", "application/json")
                .build();
    }
}
